[WIP] Add prototype of new data registry package

ellisonbg opened the pull request:
Fixes #5548 

This pull request contains an initial prototype of a database for JupyterLab. The current implementation is fairly simple, but covers the main points. 

To do:

- [ ] Write tests
- [ ] Review design
- [x] Add logic to test for already published datasets
- [ ] Create databus extension to make this available to other extensions.

ellisonbg commented:
@ian-r-rose @saulshanabrook

saulshanabrook commented on packages/databus/src/index.ts:
```diff
@@ -0,0 +1,133 @@
+    this._datasetsChanged.emit({ dataset, type: 'added' });
+
+    return new DisposableDelegate(() => {
+      ArrayExt.removeAt(this._datasets, index);
```
I think the index might not necessarily be valid when it is removed. For example, let's say A is added, then B is added. So at adding time, B is at index 1. Then A is removed. Now, B is at index 0. But, the index here will still be 1.

ellisonbg replied:
Good catch, push fix.

saulshanabrook commented on packages/databus/src/index.ts:
```diff
@@ -0,0 +1,133 @@
+   * We are currently exploring different metadata schemes, so this may change
+   * in the future.
+   */
+  metadata: any;
```
Until we have some sort of typing for metadata, it might clearer not to include it or type it as an empty object. Otherwise, as a user, I am not sure to tell whether to put some part of the data in the metadata or in the actual data.

ellisonbg replied:
Yeah, until we figure out more about the metadata, we may need to hedge a bit.

saulshanabrook replied:
Reading your comments on the issue, it seems like one way to proceed would be to say "A certain mimetype corresponds to a certain structure of data/metadata".

If we were thinking about this problem in Typescript land, that sounds a lot like a subclass. So could have a `Data` abstract base class, that requires mimetype, metadata and data properties, but metadata and data are generically typed.  Then you would subclass from that for CSV data, to narrow down possibilities of what is in those fields. [Here is an example](http://www.typescriptlang.org/play/#src=interface%20IData%3CT%2C%20V%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20String%0D%0A%20%20%20%20data%3A%20T%0D%0A%20%20%20%20metadata%3A%20V%0D%0A%7D%0D%0A%0D%0A%0D%0Ainterface%20ICSVData%3CT%3E%20extends%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20Array%3CT%3E%2C%0D%0A%20%20%20%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20c%3A%20ICSVData%20%3C%20%7B%20name%3A%20string%20%7D%20%3E%20%3D%20%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%0D%0A%20%20%20%20metadata%3A%20%7B%22name%22%3A%20%22string%22%7D%0D%0A%7D%0D%0A%0D%0A%0D%0A%2F%2F%2F%20or%20could%20do%20class%20based%20approach%0D%0A%0D%0Aclass%20CSVData%3CT%3E%20implements%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%20%3D%20%22CSV%22%3B%0D%0A%0D%0A%20%20%20%20constructor(public%20data%3A%20Array%3CT%3E%2C%20public%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D)%20%7B%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%0D%0Aconst%20c2%3A%20CSVData%3C%7B%20name%3A%20string%20%7D%3E%20%3D%20new%20CSVData(%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%20%7B%20%22name%22%3A%20%22string%22%20%7D)%3B)

ellisonbg commented:
+1

On Thu, Dec 6, 2018 at 11:07 AM Saul Shanabrook <notifications@github.com>
wrote:

> *@saulshanabrook* commented on this pull request.
> ------------------------------
>
> In packages/databus/src/index.ts
> <https://github.com/jupyterlab/jupyterlab/pull/5734#discussion_r239575963>
> :
>
> > +   * #### Notes
> +   * The notion of data in the databus is completely abstract, and typed as an
> +   * any. Publishers and consumers of a given mimetype will need to agree
> +   * what the actual data is and cast it appropriately before use.
> +   */
> +  data: any;
> +
> +  /**
> +   * The metadata associate with the dataset.
> +   *
> +   * #### Notes
> +   * Like the data itself, the databus notion of metadata is entirely abstract.
> +   * We are currently exploring different metadata schemes, so this may change
> +   * in the future.
> +   */
> +  metadata: any;
>
> Reading your comments on the issue, it seems like one way to proceed would
> be to say "A certain mimetype corresponds to a certain structure of
> data/metadata".
>
> If we were thinking about this problem in Typescript land, that sounds a
> lot like a subclass. So could have a Data abstract base class, that
> requires mimetype, metadata and data properties, but metadata and data are
> generically typed. Then you would subclass from that for CSV data, to
> narrow down possibilities of what is in those fields. Here is an example
> <http://www.typescriptlang.org/play/#src=interface%20IData%3CT%2C%20V%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20String%0D%0A%20%20%20%20data%3A%20T%0D%0A%20%20%20%20metadata%3A%20V%0D%0A%7D%0D%0A%0D%0A%0D%0Ainterface%20ICSVData%3CT%3E%20extends%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20Array%3CT%3E%2C%0D%0A%20%20%20%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20c%3A%20ICSVData%20%3C%20%7B%20name%3A%20string%20%7D%20%3E%20%3D%20%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%0D%0A%20%20%20%20metadata%3A%20%7B%22name%22%3A%20%22string%22%7D%0D%0A%7D%0D%0A%0D%0A%0D%0A%2F%2F%2F%20or%20could%20do%20class%20based%20approach%0D%0A%0D%0Aclass%20CSVData%3CT%3E%20implements%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%20%3D%20%22CSV%22%3B%0D%0A%0D%0A%20%20%20%20constructor(public%20data%3A%20Array%3CT%3E%2C%20public%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D)%20%7B%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%0D%0Aconst%20c2%3A%20CSVData%3C%7B%20name%3A%20string%20%7D%3E%20%3D%20new%20CSVData(%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%20%7B%20%22name%22%3A%20%22string%22%20%7D)%3B>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyterlab/jupyterlab/pull/5734#discussion_r239575963>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABr0AddemhMBk_UJEGQdYobGpFHiIF0ks5u2WrYgaJpZM4ZHEww>
> .
>


-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

ellisonbg commented:
@saulshanabrook see the latest iteration for an attempt to type the data/metadata. Any ideas on avoiding the `IDataset<any,any>` in the `DataBus`?

saulshanabrook commented:
> Any ideas on avoiding the IDataset<any,any> in the DataBus?


I think we have to have the list be that, but I think we can provide a typed `filter` function, that instead of taking in a mimetype directly it takes in a function which returns whether a dataset is of a certain type. You can use Typescript's [user defined type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) to make this type OK.

See [this example](http://www.typescriptlang.org/play/#src=interface%20IData%3CT%2C%20V%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20String%0D%0A%20%20%20%20data%3A%20T%0D%0A%20%20%20%20metadata%3A%20V%0D%0A%7D%0D%0A%0D%0A%0D%0Aconst%20l%3A%20Array%3CIData%3Cany%2C%20any%3E%3E%20%3D%20%5B%5D%3B%20%0D%0A%0D%0A%0D%0A%0D%0A%0D%0Ainterface%20ICSVData%3CT%3E%20extends%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20Array%3CT%3E%2C%0D%0A%20%20%20%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20c%3A%20ICSVData%20%3C%20%7B%20name%3A%20string%20%7D%20%3E%20%3D%20%20%7B%0D%0A%20%20%20%20mimeType%3A%20%22CSV%22%2C%0D%0A%20%20%20%20data%3A%20%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%0D%0A%20%20%20%20metadata%3A%20%7B%22name%22%3A%20%22string%22%7D%0D%0A%7D%0D%0A%0D%0A%0D%0Afunction%20isCSVData%3CT%3E(d%3A%20IData%3Cany%2C%20any%3E)%3A%20d%20is%20ICSVData%20%3C%20T%20%3E%20%7B%0D%0A%20%20%20%20if%20(d.mimeType%20%3D%3D%20%22CSV%22)%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20true%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20return%20false%3B%0D%0A%7D%0D%0A%0D%0Afunction%20retrieve%3CT%20extends%20IData%3Cany%2C%20any%3E%3E(filter%3A%20(d%3A%20IData%3Cany%2C%20any%3E)%20%3D%3E%20d%20is%20T)%3A%20Array%3CT%3E%20%7B%0D%0A%20%20%20%20const%20res%3A%20Array%3CT%3E%20%3D%20%5B%5D%3B%0D%0A%20%20%20%20for%20(let%20d%20of%20l)%20%7B%0D%0A%20%20%20%20%20%20%20%20if%20(filter(d))%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20res.push(d)%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20return%20res%3B%0D%0A%7D%0D%0A%0D%0A%0D%0Al.push(%7B%20mimeType%3A%20%22CSV%22%2C%20data%3A%20null%2C%20metadata%3A%20null%20%7D)%3B%0D%0Al.push(%7B%20mimeType%3A%20%22other%22%2C%20data%3A%20null%2C%20metadata%3A%20null%20%7D)%3B%0D%0A%0D%0Aconst%20allCSVData%20%3D%20retrieve(isCSVData)%3B%0D%0Aconsole.log(allCSVData)%3B%0D%0A%0D%0A%0D%0A%2F%2F%2F%20or%20could%20do%20class%20based%20approach%0D%0A%0D%0Aclass%20CSVData%3CT%3E%20implements%20IData%3CArray%3CT%3E%2C%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D%3E%20%7B%0D%0A%20%20%20%20mimeType%20%3D%20%22CSV%22%3B%0D%0A%0D%0A%20%20%20%20constructor(public%20data%3A%20Array%3CT%3E%2C%20public%20metadata%3A%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20string%20%7D)%20%7B%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20isCSVDataClass%3CT%3E(d%3A%20IData%3Cany%2C%20any%3E)%3A%20d%20is%20CSVData%3CT%3E%20%7B%0D%0A%20%20%20%20return%20d%20instanceof%20CSVData%3B%0D%0A%7D%0D%0A%0D%0Al.push(new%20CSVData(%5B%7B%20name%3A%20%22hi%22%20%7D%5D%2C%20%7B%20%22name%22%3A%20%22string%22%20%7D))%3B%0D%0A%0D%0Aconsole.log(retrieve(isCSVDataClass))%3B%0D%0A) for how I added a `retrieve` function that only returns a particular kind of data.

This would let users put types in and get types out, as long as their type guard functions were right.

As you can see in the example, you can write type guards both for plain objects and for classes.

saulshanabrook commented:
I am curious if we should require `mimetype`, `data` and `metadata` or if we can let users defined datasets however they like.

I think to answer that, we have to flesh out more how the databus will be used internally and externally.

ian-r-rose commented on packages/databus/src/index.ts:
```diff
@@ -0,0 +1,135 @@
+/**
+ * A databus object, for managing datasets in JupyterLab.
+ */
+export class DataBus {
```
What did you think of the discussion with @BoPeng about the naming of this? That is, we aren't really talking about shuttling data around, so much as registering and publishing it.

BoPeng replied:
Perhaps we can call items to be managed as `DataToken`, and managers in line of `DataRegistry` and `DataPublisher`?

ellisonbg replied:
I am completely open to other names. I think `DataRegistry` is probably pretty neutral and descriptive.

ian-r-rose commented on packages/databus/src/index.ts:
```diff
@@ -0,0 +1,135 @@
+   *
+   * @returns An array of matching `IDataset` objects.
+   */
+  filter(mimeType: string): Array<IDataset<any, any>> {
```
I like this. It has a flavor of the `IInstanceTracker`.

ian-r-rose replied:
Though, I wonder if `filter` could be more general and take a function `item => bool`. We could then filter based on mimetype globs, for instance (`'test/*'`).

ellisonbg replied:
I like the idea of taking a function. It is strongly typed, it makes it easy to also filter on details of metadata.

ian-r-rose commented on packages/databus/src/index.ts:
```diff
@@ -0,0 +1,135 @@
+  /**
+   * Return a list of all published datasets.
+   */
+  get datasets(): Array<IDataset<any, any>> {
```
Sometimes with a datastructure like this we expose an iterator, somtimes an array. I don't think I have a *super* strong gut feeling for when we should choose one over the other, but an argument for an iterator could be

a. when the number of datasets could be dynamic or unknown, and
b. when we want to emphasize that the ordering is meaningless and should not be relied upon.

What would you say to making this an iterator instead?

saulshanabrook replied:
Another option would be to make this an [iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators#User-defined_iterables), which would allow users to loop through it or turn it into an array if they would like (with `[...databus]`).

ellisonbg replied:
Don't feel strongly about using Array, but I do think it is important to emphasize the the order should not be relied upon. I can look into the iterator and iterable.

ian-r-rose commented on packages/databus/src/index.ts:
```diff
@@ -0,0 +1,135 @@
+
+    this._datasetsChanged.emit({ dataset, type: 'added' });
+
+    return new DisposableDelegate(() => {
```
A consequence of this pattern is that only the publisher of a dataset will be able to remove it. I think that makes good sense, just trying to think through the lifecycle a bit...

ellisonbg replied:
Yes, I thought of that as well. I can imagine situations where others might want to remove a dataset, but I also don't want to over engineer it. Ideas on an alterative?

saulshanabrook commented on packages/databus/src/index.ts:
```diff
@@ -0,0 +1,135 @@
+    return this._datasetsChanged;
+  }
+
+  private _datasets: Array<IDataset> = [];
```
It might make sense to make this a [Set](https://howtodoinjava.com/typescript/sets/) instead of an array, because we don't allow duplicate values and we don't care about order.

ellisonbg replied:
+1

ian-r-rose commented:
I like the idea of templating as you and @saulshanabrook have worked out.

I am also trying to think through some specific use-cases for how other plugins might use this. That is to say: we expose an `IDataBus` token, which may then be required by other plugins. What plugins might want to publish datasets? A few thoughts:

1. A large-remote-dataset-aware filebrowser could send an `IDataSet` instead of the full contents, and the filebrowser-plugin could register this.
2. An IRODS or Intake catalogue viewer could register a dataset.
3. An Arrow client could register a remote dataframe.

On the other end, what could consumers then do with the `IDataBus`?

1. Pop open a preview or metadata pane in the main area.
2. Get some drag/drop code-snippets for opening, inspecting, closing the dataset in a notebook.
3. Provide some sort of commenting/annotation system.
4. ...?

Overall I see this as trying to provide the smoothest ramp to interacting with the data in a proper kernel.

saulshanabrook commented:
@ian-r-rose 

> What plugins might want to publish datasets?

Another thing to think about is that plugins which publish datasets probably also would expose their own interface or class which extends `IDataSet` and a filter function. That way, other plugins which want to use their datasets can filter for that specific type and understand it's particular semantics.

ian-r-rose commented:
@saulshanabrook Agreed.

ellisonbg commented:
Other potential publishers of data:

* Certain mime renderers that get actual data that could be of use of other extensions. This would cover both mime renderers in the notebook and for files.
* A SQL extension that talks to a remote SQL DB and registers the tables in that DB.

ellisonbg commented:
I have done a bunch of work and updated the APIs:

* `IDataset<A>` now has an optional URI, but no metadata (that will set elsewhere and the URI will be used for pointing.
* `DataBus` is now `DataRegistery`.
* A `Set` is used to store datasets.
* Added a converter module that has a registry of converters between different dataset types.

saulshanabrook commented on packages/databus/src/dataregistry.ts:
```diff
@@ -0,0 +1,136 @@
+   *
+   * @returns An set of matching `IDataset` objects.
+   */
+  filter<T>(func: (value: IDataset<T>) => boolean): Set<IDataset<T>> {
```
```suggestion
  filter<T extends IDataset<any>>(func: (value: IDataset<any>) => value is T): Set<T> {
```

saulshanabrook replied:
These suggestions change the filter function to take a [user defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) instead of just a function that returns a boolean. This allows us to skip the `as IDataset<T>` later on, because Typescript knows that if this function returns `True`, then `value` is of type  `T`.

saulshanabrook commented on packages/databus/src/dataregistry.ts:
```diff
@@ -0,0 +1,136 @@
+   * @returns An set of matching `IDataset` objects.
+   */
+  filter<T>(func: (value: IDataset<T>) => boolean): Set<IDataset<T>> {
+    let result: Set<IDataset<T>> = new Set();
```
```suggestion
    let result: Set<T> = new Set();
```

saulshanabrook commented on packages/databus/src/dataregistry.ts:
```diff
@@ -0,0 +1,136 @@
+   */
+  filter<T>(func: (value: IDataset<T>) => boolean): Set<IDataset<T>> {
+    let result: Set<IDataset<T>> = new Set();
+    this._datasets.forEach((value: IDataset<T>) => {
```
```suggestion
    this._datasets.forEach((value: T) => {
```

saulshanabrook commented on packages/databus/src/dataregistry.ts:
```diff
@@ -0,0 +1,136 @@
+  filter<T>(func: (value: IDataset<T>) => boolean): Set<IDataset<T>> {
+    let result: Set<IDataset<T>> = new Set();
+    this._datasets.forEach((value: IDataset<T>) => {
+      if (func(value as IDataset<T>)) {
```
```suggestion
      if (func(value)) {
```

saulshanabrook commented:
I see three options for where this is worked on:

* **Brian's branch**: current spot, works well when Brian is only one working on it. Hard for others to make PRs against
* **JupyterLab master branch**: Easy for people to create PRs against. However, then is linked to JupyterLab release cycle. This might be an issue if we don't want to release this work but do want to have a jupyterlab release.
* **New JupyterLab branch**: Easy to make PRs against, need to keep up to date with master.
* **Separate repo**: Easy to make PRs against, but means that other parts of JupyterLab core can't depend on it

@ellisonbg @ian-r-rose  Do you all have thoughts? Seems like choice is between master and new branch in jupyterlab repo. I am leaning towards new branch, so that we don't have to release before we want to.

saulshanabrook commented:
@ellisonbg I just opened this PR against this branch https://github.com/ellisonbg/jupyterlab/pull/4 

* Removes `data` attribute of `IDataset`. Why? We isn't needed currently because we don't actually depend on the `data` attribute for anything. If you have some type of dataset that holds data in some way, just write an interface or class that implements the `IDataSet` interface and hold your data wherever you want to.
* Removes empty constructor of `DataRegistry`
* Make `DataRegistry.filter` take in a predicate function so casting isn't needed
* Change `ConvertRegistry._converters` to be `Map` of `Set`. instead of object of arrays. Why? `Map` seems to be closer to the semantic representation of what we are doing, i.e. mapping data converter sources to the data converters. Also, I don't believe having duplicate `IConverter`s is useful so a set seems suitable there.
* Removes empty constructor fo `ConverterRegistry`
* Switches `listTargetMimeTypes` to return a `Set` instead of an `Array`. Why? I don't think it's useful to return duplicate mime types.

saulshanabrook commented:
TODO:

- [ ] Add branch on JupyterLab core and open PR with that branch. Close this PR.
- [ ] Update my PR to remove data change
- [ ] Merge my PR into new branch.

saulshanabrook commented:
I am closing this in favor of https://github.com/jupyterlab/jupyterlab/pull/5857, which extends it but on a branch on the core repo.
